README.md
<p align="center">
  <img src="https://raw.githubusercontent.com/PureFunctor/amalgam-lisp/develop/docs/logo.png"></img>
</p>

<p align="center">
  LISP-like interpreted language implemented in Python.
</p>

<p align="center">
  <a href="https://github.com/PureFunctor/amalgam-lisp/actions">
    <img src="https://img.shields.io/github/workflow/status/PureFunctor/amalgam-lisp/Tests?logo=github&style=flat-square">
    </a>
  <a href="https://codecov.io/gh/PureFunctor/amalgam-lisp/">
    <img src="https://img.shields.io/codecov/c/gh/purefunctor/amalgam-lisp?label=codecov&logo=codecov&style=flat-square">
  </a>
  <a href="https://amalgam-lisp.readthedocs.io/">
    <img src="https://img.shields.io/readthedocs/amalgam-lisp?style=flat-square">
  </a>
  <a href="https://pypi.org/project/amalgam-lisp/">
    <img src="https://img.shields.io/pypi/v/amalgam-lisp?style=flat-square">
  </a>
  <a href="https://pypi.org/project/amalgam-lisp/">
    <img src="https://img.shields.io/pypi/pyversions/amalgam-lisp?style=flat-square" >
  </a>
</p>

# Installation & Basic Usage
This package can be installed through PyPI:
```bash
$ pip install amalgam-lisp
```
This makes the `amalgam` command-line script available.
```bash
$ amalgam                     # To invoke the REPL
$ amalgam hello.am            # To load and run a file
$ amalgam --expr="(+ 42 42)"  # To evaluate an expression
```

# Development Setup
Install the following dependencies:
* Python 3.7 & 3.8
* [Poetry](https://python-poetry.org)
* [Nox](https://nox.thea.codes/en/stable/) (Optional)

Clone and then navigate to the repository:
```bash
$ git clone https://github.com/PureFunctor/amalgam-lisp.git
$ cd amalgam-lisp
```

Install the dependencies for the project:
```bash
$ poetry install
$ poetry run pre-commit install
```

## Running Tests / Coverage Reports / Building Documentation
This project uses `nox` for automating various tasks like running tests and building documentation.
```
$ nox
```

Alternatively, tests, coverage reports, and the documentation can be generated manually.
```bash
$ poetry run coverage run -m pytest
$ poetry run coverage report -m
$ poetry run coverage html
$ poetry run sphinx-build docs docs/build
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
* `__iter__` and `__len__` for `SExpression` and `Vector`.
* Special `&rest` syntax for `fn`, `mkfn`, and `macro` through `createfn`.
* `AmalgamMeta` to be used by `Amalgam` instead of inheriting from `ABC`.
* `Failure` and `FailureStack` to implement a notification framework through `AmalgamMeta`.

### Changed
* Manually handle uncallable types in `SExpression.evaluate`.
* Account for subclasses when type checking `Located.located_on`.
* Reorganize `primordials.py`, transforming it into a subpackage.
* Refactor internal code to use the new notification framework.

### Removed
* The `bind` and `call` methods from `Amalgam`, favoring manual checks instead.
* The `Internal`, `Trace`, and `Notification` classes in favor of `Failure` and `FailureStack`.


## [0.2.0] 2020-11-13
This marks the second volative release before `v1.0.0`.

### Added
* `Located` mixin dataclass for AST nodes.
* `__repr__` for `Environment`.
* `name` and `engine` parameters and attributes to `Engine`.
* `_interpret` and `interpret` methods to `Engine`.
* `Notification` class for representing failures.
* Tracking of the original text and its source during parsing and evaluation.

### Changed
* Unquote unevaluated arguments in deferred functions and macros.
* Update user-facing documentation for the `macro` function.
* Move parse buffer and continuation logic to `Engine.repl`.
* Visualize `Notification`s in `Engine.repl`.
* Use `Notification`s in `_loop`, `_return`, and `_break`.

### Fixed
* Unnecessary imports for type checking.

### Removed
* The `_require` and `_provide` built-in functions.
* The `DisallowedContextError` exception.
* The `SymbolNotFound` exception.
* The `parse_and_run` method of `Engine`.
* The `Parser` class in favor of the `parse` function.
* Type annotations for the transformer methods in `Expression`.


## [0.1.0] 2020-11-2
This marks the first volatile feature release before `v1.0.0`.

### Added
* Implementations and tests for the following:
  * Language entities: `amalgam.amalgams`.
  * Execution environments: `amalgam.environment`.
  * Runtime engine: `amalgam.engine`.
  * Built-in functions: `amalgam.primordials`.
  * Parser built with Lark: `amalgam.parser`.
  * Command-line interface: `amalgam.cli`.
* User-facing and internal documentation.
* Type annotaions and runtime casts for `mypy`.


noxfile.py
import tempfile

import nox


def install_with_constraints(session, *args, **kwargs):
    """
    https://cjolowicz.github.io/posts/hypermodern-python-03-linting/
    """
    with tempfile.NamedTemporaryFile() as requirements:
        session.run(
            "poetry",
            "export",
            "--dev",
            "--format=requirements.txt",
            f"--output={requirements.name}",
            external=True,
        )
        session.install(f"--constraint={requirements.name}", *args, **kwargs)


@nox.session
def docs(session):
    install_with_constraints(session, "sphinx", "sphinx_rtd_theme")
    session.install(".")
    session.run("sphinx-build", "docs", "docs/build")


@nox.session
def lint(session):
    install_with_constraints(session, "pre-commit")
    session.run("pre-commit", "install")
    session.run("pre-commit", "run", "--all-files")


@nox.session(python=("3.7", "3.8", "3.9"))
def test(session):
    env = {"COVERAGE_FILE": f".coverage.{session.python}"}

    install_with_constraints(session, "pytest", "pytest-mock", "coverage")
    session.install(".")
    session.run("coverage", "run", "--branch", "-m", "pytest", "-vs", env=env)
    session.run("coverage", "report", "-m", env=env)


@nox.session
def coverage(session):
    install_with_constraints(session, "coverage[toml]")
    session.run("coverage", "combine")
    session.run("coverage", "report", "-m")
    session.run("coverage", "html")
